### DISCUSSION TITLE
ATTiny1634 and Servos

### DISCUSSION THREAD

**Discussion by triodit** [SpenceKonde/ATTinyCore#588] [Library discussion]
Hi guys Total newbling here trying to learn all the things.  I'm working on an ambitions project and I'm trying to drive 4 servos on a 1634.  So far I've managed to drive 0 servos cause I can't figure out what libraries to use :(    can anyone help a dumb dumb figure this out?

**Comment by SpenceKonde** [comment]
The 1634 has a normal Timer1. The normal implementation in Servo.h works 

However, if you have installed the Servo library manually (such that there's a (sketchbook)/libraries/Servo/Servo.h file), including Servo.h will use that one, not the version included by the core. This is generally bad - I don't know if the "stock
 version of Servo.h is written generically enough that it will realize it can just use the default implementation (that's often a problem with Arduino libraries (the official developers are awful about that, and were even worse in the past). If you end up with it pulling in the wrong version of Servo.h, I *also* include an identical copy of that library differing only in name, Servo_ATTinyCore - Everything about the way the library works is otherwise identical to the standard one, you just #include <Servo_ATTinyCore.h> instead of <Servo.h> 
 
 If you're getting compile errors, post the text of the errors. 
 
 If it compiles and uploads but misbehaves, as with any motor, you should  be suspicious of possible power supply problems (either the power source not being able to quite supply enough current for the inductive load, or the electrical noise that results from it being turned on and off); the most common manifestation of this that you'd see is that the sketch gets to where it tries to move the servo, you hear a bit of noise from the servo and then the sketch restarts.

**Reply by triodit** [reply]
Ok that compiles, now to figure out why I'm not seeing the servo moving.......  hmmmmm   I'm not worried about the power, I designed the board with a stupidly overkill power system.  but ya hmmmmm not seeing PWM on the lines with my o-scope..  I'll have to look at it agian in the morning.  But adding _ATTinyCore helped a ton :D  don't mind me I'm jsut slow.

**Comment by SpenceKonde** [comment]
You've got a 'scope! Thank god, someone else has a properly outfitted
workshop...

Okay, this shouldn't be too hard since you have a scope in front of you
instead of blinders across your eyes.

Soo- first sanity check the same code on a nano or uno or something like
that just to make sure you're not doing somethingsilly.

If that works but 1634 doesn't, I'll give you more diagnostic code. It is
entirely possible that you're the first person using this on 1634, or the
first person who is telling me when they have a problem instead of
shrugging their shoulders and assuming it is known issue. (A classic tech
person rant is about people who ask when something will be fixed, when the
person being asked, and whose job is to fix it was never told it was
broken) there certainly 8snt any fundamental reason it should have issues
on 1634. It's probably something really dumb in my code like setting the
correct bit b in the wrong register or something because I didn't realize
it was sometimes elsewhere. Timer1 is identical among almost all the
classic avrs
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by triodit** [reply]
I learned to do all this backwards xD.  I'm a master electronics tech/fabricator trying to learn to design and code.  so I know what signals look like from years of trouble shooting.  and IDK how ANYONE would learn this stuff if they couldn't look at the signals.  
![20210705_150021](https://user-images.githubusercontent.com/82182510/124520734-26c7c980-dda2-11eb-936a-ae876d668875.jpg)

**Comment by triodit** [comment]
Ya I'm stupped,  Heres my test code,  worked fine on a Pico Pi, and an uno.  but I'll bee dipped in sheep shit if I can figure out what I'm doing wronge.  I've tried ever possible pin assignment, 0-20, A0-A11, PA5, PA6, PB3 and PC0.  I tried scanning in groups of 5 then one at a time.  I even thought it might be hardware and I broke out a fresh ATTINY1634.  but no Pins ever put out PWM.  

#include <Servo_ATTinyCore.h>
 
int main()
{
    Servo myServo1;
//    Servo myServo2;
//    Servo myServo3;
//    Servo myServo4;
//    Servo myServo5;
    
    
                        //a maximum of five servo objects can be created 
 
    myServo1.attach(11);  //attach the servo to pin PB1
//    myServo2.attach(A9);  //attach the servo to pin PB1
//    myServo3.attach(A7);  //attach the servo to pin PB1
//    myServo4.attach(A8);  //attach the servo to pin PB1
//    myServo5.attach(A9);  //attach the servo to pin PB1

    //sweep the servo
    while(1)
    {
        for(int pos = 0; pos < 180; pos++)  // goes from 0 degrees to 180 degrees
        {                                   // in steps of 1 degree
            myServo1.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo2.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo3.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo4.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo5.write(pos);             // tell servo to go to position in variable 'pos'


            delay(15);                      // waits 15ms for the servo to reach the position
        }
 
        for(int pos = 180; pos > 1; pos--)  // goes from 180 degrees to 0 degrees
        {
            myServo1.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo2.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo3.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo4.write(pos);             // tell servo to go to position in variable 'pos'
//            myServo5.write(pos);             // tell servo to go to position in variable 'pos'

            delay(15);                      // waits 15ms for the servo to reach the position
        }
    }
}

**Comment by triodit** [comment]
Ok the below worked.  HOW!


/* Sweep
 by BARRAGAN <http://barraganstudio.com>
 This example code is in the public domain.

 modified 8 Nov 2013
 by Scott Fitzgerald
 http://www.arduino.cc/en/Tutorial/Sweep
*/

#include <Servo_ATTinyCore.h>

Servo myservo;  // create servo object to control a servo
// twelve servo objects can be created on most boards

int pos = 0;    // variable to store the servo position

void setup() {
  myservo.attach(14);  // attaches the servo on pin 9 to the servo object
}

void loop() {
  for (pos = 0; pos <= 180; pos += 1) { // goes from 0 degrees to 180 degrees
    // in steps of 1 degree
    myservo.write(pos);              // tell servo to go to position in variable 'pos'
    delay(15);                       // waits 15ms for the servo to reach the position
  }
  for (pos = 180; pos >= 0; pos -= 1) { // goes from 180 degrees to 0 degrees
    myservo.write(pos);              // tell servo to go to position in variable 'pos'
    delay(15);                       // waits 15ms for the servo to reach the position
  }
}

**Comment by SpenceKonde** [comment]
The hell..... 

Analog pin numbers aren't expected to work (right now the analog pin number system on ATTinyCore is a shitshow - it isn't even consistent between parts, and with anything other than analogRead() behavior is often wrong and never detected. But aside from that, both PIN _Pxn and the numbers ought to work. (a large amount of the effort that has gone towards 2.0.0 (in the 2.0.0-dev branch) is aimed at dealing with the Analog Folly - the next step is that I need deal with boards.txt, which was and still is an absolute mess - only now it's not even a valid mess because of the other changes I've made. I was hoping to get 2.0.0 out long ago, but megaTinyCore and DxCores are taking up a lot of my time these days. 

.... 

....

Oh.... 

don't override main(). If you do override main(), if you call any Arduino API functions or use any libraries written for Arduino, you **must** first call init(). (the stock main() calls the internal initialization function init(), and then the user initialization function, setup(), and then calls loop() in an infinite loop).  Functions that are part of the Arduino API assume that the part is configured in a certain way: analogRead() assumes that the ADC is enabled and set to run at a valid ADC clock [I think if it's not enabled, it straight up hangs, because it gets to the part where it loops until the ADC conversion complete flag is set, and that never happens because the ADC isn't enabled]. analogWrite() assumes that all timers that can run PWM have been kicked off, and are running at a frequency appropriate for PWM.... [if they're not running, it won't start them] and delay() and millis() assume that timer0 is running, and firing an interrupt periodically to update the running total number of elapsed milliseconds [delay() will hang waiting for 1000 microseconds to pass, but without the millis timer running, that never happens - millis just returns 0 every time. ]. 

One of my cores got into trouble once because of an assumption like that that was violated - class constructors will run before init(), and in  the case of the part in question, it did analogWrite() in a pin in the constructor. Which assumes the timer is running, and because it's a wacky timer on that part, not a sane one (it's the successor to those async timers on the x5 and x61) except they finbally made the synchronization rules explicit). So it has to disable the timer, turn PWM on, then reenable it. But it wasn't enabled to start with, so it would get enabled then, with the default prescaler (ie, no prescaling, instead of the usual div by 32). Aaaand that was also the millis timer. So when init() ran afterwards, it blindly tried to configure the timer, couldn't (because it's async, you can't reconfigure without disabling it first) and then let millis run at 32x speed. 

Libraries make similar assumptions - I do not know that Servo will initialize Timer1 from it's reset state (that is not within it's design parameters - it is only required to reconfigure it from being a PWM controller to servo controller). It probably *shouldn't* work if init() hasn't been called, because that's gotta be at least a register write or two that isn't necessary, probably burning 6 bytes each. (instruction words are 16 bits, and most special function registers require the two-word STS instruction to write, plus the Load Immediate to put the value to be written into a working register. For people using 2k parts, 3 extra random register writes like that is almost 1% of the flash for them. 

It was an absolute  bonanza when I found, during 1.50 dev, line after line after line of inappropriate use of sbi()and cbi() macros - on the first 32 special function registers (the so-called "low I/O space") , they generate a 1 word set/clear bit index instruction. On the next 32, it generates 3 word read modify write sequences and on everything else, 5 word ones.... and the core had those used repeatedly on the same registers, during initialization (when they knew, or should have known, what the initial state of the register was, so they didn't have to set most of them at all, and could do the whole thing with a single, simple assignment.to each register instead. Or maybe that was 1.4.x dev, I think that's what it was. But that;s the first time I was looking back at it after I had been in megaTinyCore land for a while, and had gotten a lot better at AVR.
(why only64 total "I/O" registers that can be more easily accessed? They didn't have enough space in the 16-bit instruction set for more: 1/16th of the instruction set is the single cycle read-write instructions, since they can read or write (2) any of the 64 locations in the I/O space to or from any of the 32 working registers 32 x 64 = 2048 out of the 65536 possible 16-bit instructions each each! There are another 10 instructions which each eat up 1/16th of the possible instructions, leaving around 20k instructions for everyrhing else. including ,math, other load-store, flow control, sbi/cbi ,etc. The top 10 instructions are literally 1/16th of the instruction space each, and the next four are 1/32nd each. there's 3/4ths of the possible opcodes right there. Until I made a chart because I had to modify a hex file without the source, I had no idea how uneven it was.

**Reply by triodit** [reply]
Oh wow I'm gonna have to read that one slowly.  Not complaining mind you but wow.  Your bloody smart Spence.
